### DISCUSSION TITLE
n00b here with a Solax X3 gen4

### DISCUSSION THREAD

**Discussion by Zeppen** [wills106/homeassistant-solax-modbus#5] [Ideas]
Hi there!
Any idea how to make the Gen.4 of solax x3 work with this custom component?
If adding as g3 all entities end up as unknown.
Might be because of my knowledge but anyway..
Connected over ethernet to my router.
I do have a waveshare rs485 to eth but unsure how to connect it.

**Comment by infradom** [comment]
I also hava a Gen4 X3 Hybrid, and  I am not able to connect to port tcp/502.
I use a wired (ethernet) pocketLan dongle.
This problem blocks me from helping with the Gen4 testing/development

Any suggestions ?

Thanks,

**Reply by Zeppen** [reply]
For what I've learned, connecting to the COM-port over rtu should work, but i dont think the modbus protocol is the same as for previous generations.
I sent an email to solax support but no success in geting the protocol, one has to be a dealer to get it..

**Reply by Zeppen** [reply]
An update on this. I spoke to my dealer/installer and there is a different protocol for gen4. One can use the COM-port and pins 4,5 to connect to it over rtu.
But im pretty much stuck here.. Dealers have to sign a Non Disclosure Agreement to get the modbus protocol for gen4..

**Comment by infradom** [comment]
In the coming days, I will start working on a version for Gen4, as I only hava en X3 Gen4 model, the X1 part will not be tested ..

**Comment by wills106** [comment]
@Zeppen Thanks to @infradom there is now Gen4 support. You should be able to use your Inverter either directly with an RS485 - USB adapter or your Ethernet - RS485 adapter.

**Reply by Zeppen** [reply]
Ill give it a go :) thnx for the update!

**Comment by stringkillersa** [comment]
I have two Solax X3 Gen 4 inverters, from 2021. I am using a TCP/IP <=> RS485 adapter and it works fine. But I would like to read from both inverters. On the RS485 side there is no problem. I have set one inverter to RS485 adress 1 and the other to address 2.  But How do I set this up with the "Solax.-Modbus" integration?  I know I can setup several instances of the integration, but how to select which RS485 address? Would be nice to no needing TWO converters... Anyone knows?

**Reply by infradom** [reply]
The solution may depend on your TCP/IP to RS485 adapter. If it is capable of mapping the second inverter address to another tcp port, you could create two instances of the integration.
Which brand of converter are you using ?
And for my interest: do you get warnings about the 'blocking call' inside event loop? Maybe this warning only occurs on serial connections to a Gen4 device

**Reply by wills106** [reply]
Are your two inverters setup in a true Master / Slave as per page 54 of the X3 User Manual?

**Reply by stringkillersa** [reply]
@infradom I'm using this converter: https://www.waveshare.com/product/RS485-TO-ETH-B.htm
It might be possible to map adresses with another firmware, but I haven come that far yet...
Nope, none such warnings...
@wills106 Nope, they are both freerunning. Only one of them have a battery - and the docs says that the systems must be identical to run master/slave. Otherwise it hade been easier :)

**Reply by infradom** [reply]
I have not read the manual yet, but the leaflet of the waveshare converter claims it is possible to map devices to different UDP port. Not sure if this is supported in TCP mode. I will examine the manual.
For UDP mode, we would have to adapt our integration to support this. Since the cost is low, I might decide to buy such a converter and try it myself first ... 
But let me first read the manuals, if TCP mode port mapping would work, we do not need to change anything

**Reply by infradom** [reply]
I have read the waveshare manual, and I dont think using UDP will be the solution. 
In our current integration, the modbus device unit address is fixed to 1.
Maybe we could make this a configurable parameter for the integration, need to hear the opinion of @wills106
In any case, you would need to start two instances of the integration, one for each inverter.

**Reply by wills106** [reply]
At the moment with Modbus over TCP its defaulted to 502 but you can change that on setup (config flow).

So if you can say setup Inverter 1 to map to 502 and inverter 2 to map to 503 that should work?

**Reply by stringkillersa** [reply]
@wills106  @infradom I have also studied the manual and as far as I can see, it's not possible to map ports to different modbus addresses. A configurable option to select device adress would really be lovely.

**Reply by wills106** [reply]
Sorry I am getting TCP Port and Modbus Port(Unit) mixed up...

I can add it in as an option.
